atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/timestamp/InDbTimestampBoundStore.java

```diff
@@ -0,0 +1,241 @@
+                    currentLimit = null;
+                }
+
+                Throwable error = result.getError();
+                if (error instanceof SQLException) {
+                    throw PalantirSqlException.create((SQLException) error);
+                }
+                throw Throwables.rewrapAndThrowUncheckedException(error);
+            default:
+                throw new IllegalStateException("Unrecognized transaction status " + result.getStatus());
+        }
+    }
+
+    @Override
+    public synchronized long getUpperLimit() {
+        return runOperation(new Operation() {
+            @Override
+            public long run(Connection connection, Long oldLimit) throws SQLException {
+                if (oldLimit != null) {
+                    return oldLimit;
+                }
+
+                final long startVal = 10000;
+                createLimit(connection, startVal);
+                return startVal;
+            }
+        });
+    }
+
+    @Override
+    public synchronized void storeUpperLimit(final long limit) {
+        runOperation(new Operation() {
+            @Override
+            public long run(Connection connection, Long oldLimit) throws SQLException {
+                if (oldLimit != null) {
+                    writeLimit(connection, limit);
+                } else {
+                    createLimit(connection, limit);
+                }
+
+                return limit;
+            }
+        });
+    }
+
+    private Long readLimit(Connection connection) throws SQLException {
+        String sql = "SELECT last_allocated FROM " + prefixedTimestampTableName() + " FOR UPDATE";
+        QueryRunner run = new QueryRunner();
+        return run.query(connection, sql, new ResultSetHandler<Long>() {
+            @Override
+            public Long handle(ResultSet rs) throws SQLException {
+                if (rs.next()) {
+                    return rs.getLong("last_allocated");
+                } else {
+                    return null;
+                }
+            }
+        });
+    }
+
+    private void writeLimit(Connection connection, long limit) throws SQLException {
+        String updateTs = "UPDATE " + prefixedTimestampTableName() + " SET last_allocated = ?";
+        try (PreparedStatement statement = connection.prepareStatement(updateTs)) {
+            statement.setLong(1, limit);
+            statement.executeUpdate();
+        }
+    }
+
+    private void createLimit(Connection connection, long limit) throws SQLException {
+        QueryRunner run = new QueryRunner();
+        run.update(connection,
+                String.format("INSERT INTO %s (last_allocated) VALUES (?)", prefixedTimestampTableName()),
+                limit);
+    }
+
+    private void createTimestampTable(Connection connection) throws SQLException {
+        try (Statement statement = connection.createStatement()) {
+            if (getDbType(connection).equals(DBType.ORACLE)) {
```

Comment:
This is a bit odd in terms of asking a connection for a db type, but we'll use it for the time being.